pkg/controlplane/controller/defaultservice/default_service_controller_test.go

```diff
@@ -0,0 +1,284 @@
+			Labels:    map[string]string{"component": "apiserver", "provider": "kubernetes"},
+		}
+	}
+
+	type args struct {
+		serviceIP         net.IP
+		publicIP          net.IP
+		servicePort       int
+		serviceNodePort   int
+		publicServicePort int
+	}
+
+	createTests := []struct {
+		testName     string
+		args         args
+		expectCreate *v1.Service // nil means none expected
+	}{
+		{
+			testName: "service does not exist",
+			args: args{
+				serviceIP:         netutils.ParseIPSloppy("1.2.3.4"),
+				publicIP:          netutils.ParseIPSloppy("192.168.0.1"),
+				servicePort:       443,
+				serviceNodePort:   0,
+				publicServicePort: 6443,
+			},
+			expectCreate: &v1.Service{
+				ObjectMeta: om(serviceName),
+				Spec: v1.ServiceSpec{
+					Ports: []v1.ServicePort{
+						{Name: "https", Port: 443, Protocol: "TCP", TargetPort: intstr.FromInt(6443)},
+					},
+					Selector:        nil,
+					ClusterIP:       "1.2.3.4",
+					IPFamilyPolicy:  &singleStack,
+					SessionAffinity: v1.ServiceAffinityNone,
+					Type:            v1.ServiceTypeClusterIP,
+				},
+			},
+		},
+	}
+	for _, test := range createTests {
+		t.Run(test.testName, func(t *testing.T) {
+			fakeClient := fake.NewSimpleClientset()
+
+			controller, err := NewController(
+				test.args.serviceIP,
+				test.args.servicePort,
+				test.args.serviceNodePort,
+				test.args.publicIP,
+				test.args.publicServicePort,
+				reconcilers.NewNoneEndpointReconciler(),
+				0*time.Second,
+				reconcilers.NoneEndpointReconcilerType,
+				fakeClient)
+			if err != nil {
+				t.Fatal(err)
+			}
+			fakeClient.PrependReactor("create", "services", func(action k8stesting.Action) (bool, runtime.Object, error) {
+				create := action.(k8stesting.CreateAction)
+				controller.serviceInformer.GetIndexer().Add(create.GetObject())
+				return true, create.GetObject(), nil
+			})
+			// create initial service
+			err = controller.createOrUpdateMasterServiceIfNeeded()
+			if err != nil {
+				t.Errorf("case %q: unexpected error: %v", test.testName, err)
+			}
+			creates := []k8stesting.CreateAction{}
+			for _, action := range fakeClient.Actions() {
+				if action.GetVerb() == "create" {
+					creates = append(creates, action.(k8stesting.CreateAction))
+				}
+			}
+			if test.expectCreate != nil {
+				if len(creates) != 1 {
+					t.Errorf("case %q: unexpected creations: %v", test.testName, creates)
+				} else {
+					obj := creates[0].GetObject()
+					if e, a := test.expectCreate.ObjectMeta, obj.(*v1.Service).ObjectMeta; !reflect.DeepEqual(e, a) {
+						t.Errorf("case %q: expected create:\n%#v\ngot:\n%#v\n", test.testName, e, a)
+					}
+					if e, a := test.expectCreate.Spec, obj.(*v1.Service).Spec; !reflect.DeepEqual(e, a) {
+						t.Errorf("case %q: expected create:\n%#v\ngot:\n%#v\n", test.testName, e, a)
+					}
+				}
+			}
+			if test.expectCreate == nil && len(creates) > 1 {
+				t.Errorf("case %q: no create expected, yet saw: %v", test.testName, creates)
+			}
+		})
+	}
+}
+
```

Comment:
I'm seeing a bunch of testcases (incorrect port, incorrect type, etc.) in the old tests that you didn't copy - why? These seems useful..